retis-derive/src/lib.rs

```diff
@@ -79,13 +78,29 @@ pub fn event_type(
         .expect("Invalid tokens from event_section macro")
 }
 
-#[proc_macro_derive(EventSectionFactory)]
-pub fn derive_event_section_factory(input: TokenStream) -> TokenStream {
-    let input: DeriveInput = parse_macro_input!(input);
+#[proc_macro_attribute]
+pub fn event_section_factory(
+    args: proc_macro::TokenStream,
+    item: proc_macro::TokenStream,
+) -> proc_macro::TokenStream {
+    let input: ItemStruct = parse_macro_input!(item);
     let ident = &input.ident;
 
+    let id: syn::Expr = syn::parse(args).expect("Invalid factory id");
+
     let output = quote! {
+        #input
+
+        impl #ident {
+            pub(crate) const FACTORY_ID: u8 = #id as u8;
+
+        }
+
         impl EventSectionFactory for #ident {
+            fn factory_id(&self) -> u8 {
```

Comment:
nit: I think we can remove the "factory" part as it's redundant. The method must be called on a  factory.

Reply:
Same with `section_id()` for event sections.